config/v1/types_authentication.go

```diff
@@ -444,9 +617,16 @@ type TokenClaimValidationRule struct {
 	// +kubebuilder:default="RequiredClaim"
 	Type TokenValidationRuleType `json:"type"`
 
-	// requiredClaim allows configuring a required claim name and its expected
-	// value
+	// requiredClaim allows configuring a required claim name and its expected value
 	RequiredClaim *TokenRequiredClaim `json:"requiredClaim"`
+
+	// Expression must not exceed 4096 characters in length.
+	// Expression must not be empty.
+	// +optional
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=4096
+	Expression string `json:"expression,omitempty"`
+	Message    string `json:"message,omitempty"`
```

Comment:
Because this is a [discriminated union](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions), you'll want to add a new `TokenValidationRuleType` constant and update the `Type` field's allowed set of values in it's enum marker.

Additionally, these fields should be moved to be under another type to adhere to the discriminated union conventions.